Add resets.

hsorby opened the pull request:
Addresses issue #176 (see [here](https://docs.google.com/document/d/1MleZJF6DYp4FO4vxslAHInrRFMqrSpARa1s_dPhwYU0/edit#heading=h.z0pc1eqvm9bg) for the `reset` element specifications and [here](https://docs.google.com/document/d/1hJmyhiSY8uHlWs7tjZcZyN4RpGeUgrxU2yq1WSIDNKI/edit) for some examples).

hsorby committed:
Update object model documentation images for resets.

hsorby commented:
These are the classes that I thinking we will need to add to support resets.

hsorby commented:
In the When information element how do you distinguish between the condition math element and the value math element?

hsorby committed:
Add skeleton When class plus test.

hsorby committed:
Merge branch 'develop' into resets

nickerso commented:
As per 19.11.3 in the draft spec, the condition is the first `math` child and the value is the second `math` child.

hsorby commented:
Seems a bit undeclarative.

nickerso commented:
Standard MathML - similar to the way `piecewise` elements work...and `root`, `power`, etc.

hsorby committed:
Add Reset class.

hsorby committed:
Add reset class source files to CMake.

hsorby committed:
Add when tests for setting value and condition math elements.

Make when tests internal so that printing of the When class is not available to users.

hsorby committed:
Correct documentation for componententity.

hsorby committed:
Add reset tests.

hsorby committed:
Add algorithm header for find_if declaration.

hsorby committed:
Add the functionality to get and set variables in resets.

hsorby committed:
Remove printer_internal this won't work on Windows.

hsorby committed:
Add some tests for printing resets.

hsorby committed:
Remove ImportedEntitiy as a base class of Model

The Model class does not need to be an imported entity.  Using multiple inheritance to include ImportedEntity for Units and Components.

hsorby committed:
Merge branch 'multiple_inheritance' into resets

hsorby committed:
Update object model images in documentation.

hsorby committed:
Add reset methods to component class in object model diagram.

hsorby committed:
Add removeVariable(size_t) variant to component.

The Component class could do with a size_t variant of the removeVariable method.  This will make it consistent with the removeReset method.  Also corrected some documentation errors with removeVariable overloads.

hsorby committed:
Add basic documentation to NamedEntityImpl and EntityImpl structs.

hsorby committed:
Add comment in CMakeLists.txt.

hsorby committed:
Correct typo in src/api/libcellml/component.h.

hsorby committed:
Update rough object model entities image.

hsorby committed:
Merging in recent changes from develop.

hsorby committed:
Remove usages of Format::xml.

hsorby committed:
Merge branch 'develop' into resets

hsorby committed:
Ignore libcellml::Printer::printReset(Reset reset) const in interface printer.i.

hsorby committed:
Ignore Qt Creator autosave files.

hsorby committed:
Rename COVERAGE_TEST_AVAILABLE to COVERAGE_TESTING_AVAILABLE.  Add PREFERRED_PYTHON_VERSION to find_package calls for Python and PythonLibs to enable driving the version of Python found.

hsorby committed:
Back port the new package find module for finding Python.

This new find Python package code works on macOS unlike the PythonInterp find module currently available.

hsorby committed:
Use new find Python module scripts for finding Python executable and associated libraries.

hsorby committed:
Fix Python tests to match current implementation regarding inheritance.

hsorby committed:
Make prepare_coverage depend on python_bindings if Python bindings have been requested.

hsorby committed:
Add utility to convert an int to a string.

hsorby committed:
Update rough object model entities and overview diagrams.

hsorby committed:
Add new class OrderedEntity.

The Ordered entity class is used by both the Reset and When classes for managing the order parameter.  Add tests and python bindings for the Reset and When classes.  Improve documentation also.

hsorby committed:
Use Correct name for CMake Python_EXECUTABLE variable.

hsorby committed:
Set minimum version of CMake required to 3.2.3.

hsorby committed:
Set the minimum CMake version required to 3.5.2.

This version of CMake is the minimum version required that supports the new FindPython module.

hsorby committed:
Use Python_Interpreter_FOUND instead of PYTHONINTERP_FOUND for new find Python module.

hsorby committed:
Add a coverage test for moving whens.

hsorby committed:
Add initial skeleton code for manipulating resets in a component.

hsorby committed:
Add tests for manipulating resets in components.

MichaelClerx reviewed (commented):
Hi Hugh,
Thanks for updating the python bindings as well!

Not sure if everyone would have been willing/able to do that, so just wondering should we come up with some policy regarding this and add some notes to the contributing guidelines? Don't want to deter people from proposing fixes/improvements, but if we end up with several languages I can see it might!

agarny commented:
+1 re @MichaelClerx's suggestion. However, depending on the binding, someone might or not have the expertise to keep things in sync, so we should keep that in mind and maybe have people "in charge" of a given binding? For example, maybe @MichaelClerx for Python? In other words, although someone should try his best to keep the bindings in sync, the person in charge of a given binding would have to double check that everything is indeed "right", or get / help getting the binding in sync with the main code, if needed.

hsorby committed:
Add a test for parsing resets.

hsorby commented:
I would like to see contributors feel no pressure to do bindings that they are not comfortable with.  I would rather have the contribution no matter how small rather than not.  To that end I think it would be a good idea for there to be a person (or persons) who are responsible for an area like the bindings.

This could also be extended to something like having someone responsible for coding standard and the like.

hsorby committed:
Implement library code to pass the tests.  Improve and correct test output.

hsorby committed:
Extend the error class to handle Reset and When errors.

hsorby committed:
Add writing of errors to parser whenn parsing resets.

hsorby committed:
Remove 'const' from std:string in 'const std::vector<const std::string>' as this upsets some compilers.

hsorby committed:
Add extra check before dereferencing iterator.

hsorby committed:
Add test for checking the reset and when returned from parser error.

hsorby commented:
I think this is now ready for review.

hsorby committed:
Add documentation for resets into Python bindings component class.

hsorby committed:
Improve documentation.

hsorby committed:
Resolve warnings when running Sphinx.

MichaelClerx commented:
How mergeable is this in terms of the FindPython2,3 stuff at the top?

MichaelClerx commented:
I can't build this, because of the same issue fixed in #231

MichaelClerx commented on src/api/libcellml/orderedentity.h:
```diff
@@ -0,0 +1,79 @@
+    OrderedEntity& operator=(OrderedEntity n); /**< Assignment operator */
+
+    /**
+     * @brief Set the order for this reset.
```
...of this entity

MichaelClerx commented on src/api/libcellml/orderedentity.h:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * @brief Set the order for this reset.
+     *
+     * Sets the order for this reset.
```
entity

MichaelClerx commented on src/api/libcellml/orderedentity.h:
```diff
@@ -0,0 +1,79 @@
+    void setOrder(int order);
+
+    /**
+     * @brief Get the order for this reset.
```
entity

MichaelClerx commented on src/api/libcellml/printer.h:
```diff
@@ -160,6 +182,17 @@ class LIBCELLML_EXPORT Printer: public Logger
     std::string printEncapsulation(ComponentPtr component) const;
 
+    /**
+     * @brief Serialise the when to @c std::string.
```
Can we capitalise When or show it's not just the word "when"?

MichaelClerx commented on src/api/libcellml/reset.h:
```diff
@@ -0,0 +1,171 @@
+     * Remove the when with the given @p index. @p index must
+     * be in the range [0, \#whens).
+     *
+     * @param index The index of the when to remove.
```
again capitalising here would make it more readable :-)

MichaelClerx commented on src/api/libcellml/reset.h:
```diff
@@ -0,0 +1,171 @@
+    /**
+     * @brief Remove all whens stored in this when entity.
+     *
+     * Clears all whens that have been added to this reset.
```
all @c Whens that have been added to this @c Reset ?

MichaelClerx commented on src/api/libcellml/reset.h:
```diff
@@ -0,0 +1,171 @@
+    bool removeWhen(const WhenPtr &when);
+
+    /**
+     * @brief Remove all whens stored in this when entity.
```
in this reset, not a when

MichaelClerx commented on src/api/libcellml/reset.h:
```diff
@@ -0,0 +1,171 @@
+    void removeAllWhens();
+
+    /**
+     * @brief Tests to see if the when pointer is contained within this reset.
```
@p when etc.

MichaelClerx commented on src/api/libcellml/when.h:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @return The @c std::string MathML string of the condition.
+     */
+    std::string getCondition() const;
```
Out of interest: Will be returning objects here later?

hsorby replied:
It's a possibility :)

MichaelClerx commented on src/component.cpp:
```diff
@@ -202,5 +214,58 @@ bool Component::hasVariable(const std::string &name) const
+        reset = mPimpl->mResets.at(index);
+    }
+
+    return reset;
```
Are we happy to return null instead of using exceptions?

hsorby replied:
We are not raising any exceptions in libCellML directly.  I think it is something you avoid unless you find yourself in a truly exceptional circumstance.

MichaelClerx reviewed (commented):
Some small things in the documentation, but no other objections from me!

hsorby committed:
Update orderedentity.h

Documentation fixes.

hsorby committed:
Update printer.h

Setting classes in documentation instead of English words.

hsorby committed:
Update reset.h

Update class references in the documentation.

MichaelClerx commented:
Working through the examples now :-)
Worth noting they describe an approximation to the behaviour defined by cellml, unless we start taking limits of `dt`. Something to think about for the informative spec...

MichaelClerx commented:
Hmmmm. Would we have been better off describing these resets as changes to be made upon certain crossings? So "when B rises past 3 then A <-- 3" (like a [signal edge](https://en.wikipedia.org/wiki/Signal_edge) in electronics). It's not really about "when B > 3", because A is allowed change after that...

The "interpretation" section of the spec also has a mistake (I made a comment just now), by stating: "Each reset element describes a change to be applied to the variable referenced by the variable attribute when specified conditions are met". This is not true, the change is to be applied at the instant when a condition changes from false to true, not for the duration of the period that this transition holds!

We clarify this after, but hmmm... Maybe we need to pick our names better while we still can? Since it's confusing even ourselves

agarny commented on src/api/libcellml/component.h:
```diff
@@ -16,8 +16,9 @@ limitations under the License.
 
-#include "libcellml/componententity.h"
 #include "libcellml/exportdefinitions.h"
+#include "libcellml/componententity.h"
```
I believe we did, at some point, agree on listing include files in alphabetical order. So, we should have `componententity.h` back to where it used to be.

agarny commented on src/api/libcellml/component.h:
```diff
@@ -210,6 +213,89 @@ class LIBCELLML_EXPORT Component: public ComponentEntity
+     *
+     * @param r The reset to add.
+     */
+    void addReset(const ResetPtr &r);
```
For consistency, we might want to replace `r` with `reset` (as is the case with `removeReset()` below).

agarny commented on src/api/libcellml/componententity.h:
```diff
@@ -61,11 +61,11 @@ class LIBCELLML_EXPORT ComponentEntity: public ImportedEntity
      * @param index The index of the component to remove.
      *
-     * @return True if the units were replaced, false otherwise.
+     * @return True if the component was replaced, false otherwise.
```
"removed" rather than "replaced"?

agarny commented on src/api/libcellml/componententity.h:
```diff
@@ -82,7 +82,7 @@ class LIBCELLML_EXPORT ComponentEntity: public ImportedEntity
      * components for the component with the specified @p name. Default value is @c true.
      *
-     * @return True if the units were replaced, false otherwise.
+     * @return True if the component was replaced, false otherwise.
```
"removed" rather than "replaced"?

agarny commented on src/api/libcellml/orderedentity.h:
```diff
@@ -0,0 +1,79 @@
+    void unsetOrder();
+
+    /**
+     * @brief Test to determine of the order is set or unset.
```
"... determine if the..."

agarny commented on src/api/libcellml/reset.h:
```diff
@@ -0,0 +1,171 @@
+     * @overload
+     *
+     * @param when The pointer to the @c When to remove.
+     * @c Whens for the specified @p when pointer. Default value is @c true.
```
Not sure I understand that sentence...

agarny commented on src/api/libcellml/reset.h:
```diff
@@ -0,0 +1,171 @@
+     * @brief Tests to see if the @c When pointer is contained within this @c Reset.
+     *
+     * Tests to see if the argument @c When pointer @p when is contained
+     * within this @c When entity. Returns @c true if the @c When is in the @c Reset
```
"within this @c Reset entity."

agarny commented on src/api/libcellml/when.h:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @param condition The @c std::string MathML string of the condition to set.
+     */
+    void setCondition(const std::string& condition);
```
Minor: there should be a space before `&` and none after, i.e.

void setCondition(const std::string &condition);

The same is true in various parts of the code in fact (e.g. `setValue()` below).

agarny commented on src/bindings/interface/component.i:
```diff
@@ -70,6 +70,30 @@ other variables), this component will not be serialised in the connection.";
 "Returns the number of variables the component directly contains.";
 
+%feature("docstring") libcellml::Component::addReset
+"Add a reset `r` to this component.";
```
"Add a reset `reset` to this component.";

agarny commented on src/parser.cpp:
```diff
@@ -1125,4 +1157,216 @@ void Parser::ParserImpl::loadImport(const ImportSourcePtr &importSource, const M
+                err->setReset(reset);
+                err->setKind(Error::Kind::RESET);
+                err->setRule(SpecificationRule::RESET_INVALID_VARIABLE_REFERENCE);
+                mParser->addError(err);
```
I am not sure I like this, i.e. call `setReset()` and then `setKind()`. I mean, we could, in principle, have a call `setKind()`, but then forget to call `setReset()`. In fact, if at all possible, I would have a simple call to `mParser->addError(...);`, to which all the parameters needed to create and customise `ErrorPtr` would be passed. We could, for example, pass `reset`, but not `Error::Kind::RESET` since this would be implied. This means that we would have several signatures for `addError()`, but at least there wouldn't be any risk for errors from a library user. Just a thought...

hsorby replied:
I have done this in a later pull request (for validation #227) where I have simplified adding errors slightly.  I haven't gone the whole way on this but this can be added easily enough.

agarny replied:
Fair enough, this is not related to this PR, so it can be simplified some other time, I agree.

agarny commented on src/parser.cpp:
```diff
@@ -1125,4 +1157,216 @@ void Parser::ParserImpl::loadImport(const ImportSourcePtr &importSource, const M
+                }
+                ErrorPtr err = std::make_shared<Error>();
+                err->setDescription("Reset in component '" + component->getName() +
+                                    "' referencing variable '" + variableName +
```
We ought to handle the case where `variableName` is a null string.

hsorby replied:
Null string or empty string, I'm not clear on what you mean here.

agarny replied:
By null string, I indeed mean an empty string. `variableName` can be equal to `""`, which means that the error message would be a bit nonsensical.

agarny commented on src/parser.cpp:
```diff
@@ -1125,4 +1157,216 @@ void Parser::ParserImpl::loadImport(const ImportSourcePtr &importSource, const M
+                            "' does not have an order defined.");
+        err->setReset(reset);
+        err->setKind(Error::Kind::RESET);
+        mParser->addError(err);
```
Am I correct in assuming that we will get two error messages if a non-integer value is provided for the order? If so, is that really what we want?

hsorby replied:
Yes, I think it is what we want as we are actually violating two rules.

agarny replied:
Ok, fair enough, I don't (yet) feel too strongly about it.

agarny commented on src/parser.cpp:
```diff
@@ -1125,4 +1157,216 @@ void Parser::ParserImpl::loadImport(const ImportSourcePtr &importSource, const M
+void Parser::ParserImpl::loadWhen(const WhenPtr &when, const ResetPtr &reset, const XmlNodePtr &node)
+{
+    std::string referencedVariableName = "";
+    VariablePtr referencedVariable =  reset->getVariable();
```
(Very) minor: replace the double space before `reset` with only one space.

agarny commented on src/parser.cpp:
```diff
@@ -1125,4 +1157,216 @@ void Parser::ParserImpl::loadImport(const ImportSourcePtr &importSource, const M
+    while (childNode) {
+        if (childNode->isType("math")) {
+            // TODO: copy any namespaces declared in parents into the math element
+            //       so math is a valid subdocument.
```
Shouldn't this be done before having this PR merged in?

hsorby replied:
I'm not finding any problems with this at this point so I don't have any desire to fix something that is not broken.  If we have an example that we can test against that does cause a problem then I would fix it.

agarny replied:
Fair enough, it's just that whenever I see a `TODO`, then I would expect it implemented or deleted (i.e. have the code as clean as possible).

agarny commented on src/printer.cpp:
```diff
@@ -148,11 +150,19 @@ std::string Printer::printComponent(ComponentPtr component) const
     }
     size_t variable_count = component->variableCount();
-    if ((variable_count > 0) || (component->getMath().length())) {
+    size_t reset_count = component->resetCount();
```
Minor: for consistency, rename `variable_count` to `variableCount` and `reset_count` to `resetCount`.

agarny commented on src/utilities.cpp:
```diff
@@ -20,13 +20,15 @@ limitations under the License.
 #include <limits>
 #include <sstream>
+#include <algorithm>
+#include <set>
```
Order of includes.

agarny commented on src/utilities.cpp:
```diff
@@ -53,4 +55,41 @@ std::string convertDoubleToString(double value)
+    return validIntegerCharacters.find(c) != validIntegerCharacters.end();
+}
+
+bool convertToInt(const std::string &candidate, int *value)
```
I believe we are using C++11? If so, we might want to use [`stoi()`](http://www.cplusplus.com/reference/string/stoi/).

hsorby replied:
We are using it, we are just wrapping over it so we can do some manipulation checking if required.

agarny replied:
Oops, my bad, I didn't spot it!

agarny commented on tests/bindings/python/test_reset.py:
```diff
@@ -0,0 +1,144 @@
+        r.addWhen(w)
+
+        self.assertTrue(r.removeWhen(w))
+        self.assertEqual(0, r.whenCount())
```
What is the rationale for the 3 lines of code above? It looks like a repeat of what is above?

hsorby replied:
Different method used to remove the when one by index the other by class.

agarny replied:
Ok. Actually, it might be worth adding a comment about it?

agarny commented on tests/bindings/python/test_reset.py:
```diff
@@ -0,0 +1,144 @@
+
+        r.addWhen(w1)
+        r.addWhen(w2)
+        r.addWhen(w3)
```
Might want to add the following?

```
self.assertEqual(3, r.whenCount())
```

agarny commented on tests/coverage/coverage.cpp:
```diff
@@ -64,6 +64,24 @@ TEST(Coverage, units) {
+    wm = std::move(w);
+
+    libcellml::When wc(wm);
+
```
Delete empty line.

agarny commented on tests/parser/parser.cpp:
```diff
@@ -1417,3 +1417,177 @@ TEST(Parser, parseIds) {
+                "</component>"
+            "</model>";
+
+    const std::vector<std::string> expectedErrors = {
```
As discussed above, it might be nice to handle the case where there is no variable name (or order value) as such. Also, I can't recall whether we discussed it before (and whether it would be easy to do here), but it might be worth specifying the line/column where an error was found?

agarny commented on tests/reset/reset.cpp:
```diff
@@ -0,0 +1,467 @@
+    libcellml::WhenPtr c1 = std::make_shared<libcellml::When>();
+    libcellml::WhenPtr c2 = std::make_shared<libcellml::When>();
+    libcellml::WhenPtr c3 = std::make_shared<libcellml::When>();
+    libcellml::WhenPtr c4 = std::make_shared<libcellml::When>();
```
Delete this line since `c4` is not used.

agarny commented on tests/reset/reset.cpp:
```diff
@@ -0,0 +1,467 @@
+}
+
+TEST(Reset, printResetWithVariable) {
+    std::string e = "<reset variable=\"A\"/>";
```
`const std::string e` (and the same in a few places below).

agarny commented on tests/reset/reset.cpp:
```diff
@@ -0,0 +1,467 @@
+    libcellml::ResetPtr r = std::make_shared<libcellml::Reset>();
+    libcellml::WhenPtr w = std::make_shared<libcellml::When>();
+
+    w->setValue("<math xmlns=\"http://www.w3.org/1998/Math/MathML\">a value set</math>");
```
Am I correct in assuming that all calls to `When::setValue()` will need to include the `math` element? If anything, I would have expected a call like `w->setValue("a value set")`? I mean, the API is supposed to make our life "easy", no?

hsorby replied:
At this time I think this is fine, it wouldn't be hard to add the wrapping if not present if that is desired.

agarny replied:
Ok for now, but I will certainly like to have such a wrapper.

agarny commented on tests/when/when.cpp:
```diff
@@ -0,0 +1,57 @@
+
+    w->setCondition("<some mathml type string for value.>");
+
+    EXPECT_EQ("<some mathml type string for value.>", w->getCondition());
```
I believe we should have a call to `w->setValue()` and `w->getValue()`.

agarny commented on src/api/libcellml/componententity.h:
```diff
@@ -98,7 +98,7 @@ class LIBCELLML_EXPORT ComponentEntity: public ImportedEntity
      * components for the specified @p component pointer. Default value is @c true.
      *
-     * @return True if the units were replaced, false otherwise.
+     * @return True if the component was replaced, false otherwise.
```
"removed" rather than "replaced"?

agarny commented:
Besides my inline comments:
 - In `libCellMLRoughObjectModel-Entities.png`: why doesn't `OrderedEntity` inherit from `Entity`. The `Reset` and `When` classes are the only classes that inherit from `OrderedEntity`, and they also inherit from `Entity`.
 - In `libCellMLRoughObjectModel-Entities.png`: we are missing the `hasReset()` method in the `Component` class.
 - In `libCellMLRoughObjectModel-Entities.png`: we are missing the `isOrderSet()` method in the `OrderedEntity` class.
 - In `libCellMLRoughObjectModel-Entities.png`: we are missing the `setVariable()`, `getVariable()`, method in the `OrderedEntity` class.
 - In `libCellMLRoughObjectModel-Entities.png`: `removeWhen()` in the `Reset` class should read "removeWhen(...): void" since we have two signatures for `removeWhen()`.

Apart from that, it looks fine to me.

hsorby committed:
Update object model diagrams to show intended object model hierarchy.

hsorby committed:
Merge branch 'develop' into resets

hsorby committed:
Set OrderedEntity as a derived class of Entity, stopping When and Reset for deriving directly from Entity.

hsorby committed:
Merge branch 'resets' of github.com:hsorby/libcellml into resets

hsorby committed:
Address issues raised in review.

hsorby committed:
Fix documentation parameter names in component.h.
